databuilder/extractor/minio_stats_extractor.py

```diff
@@ -0,0 +1,113 @@
+        self.schema = 'minio'
+
+        self.s3 = boto3.resource('s3',
+                                 endpoint_url=self.endpoint_url,
+                                 aws_access_key_id=self.access_key,
+                                 aws_secret_access_key=self.secret_key,
+                                 region_name='us-east-1')
+        # TODO: use spark session from the arg parser using Spark/Kubernetes deployment
+
+    def get_stats_iter(self) -> Iterator[dict]:
+        """Get a stats iter of column stats dictionaries."""
+        stat_objs = []
+        keys = self.get_dataset_keys()
+
+        for key in keys:
+            stat_objs.extend(self.get_file_stats(key))
+
+        print("stat_objs: ", stat_objs)
+        return iter(stat_objs)
+
+    def get_dataset_keys(self) -> List[str]:
+        """Get a list of keys in an S3 bucket."""
+        dataset_paths = set()
+        result = self.s3.meta.client.list_objects(Bucket=self.bucket_name, Delimiter='/',
+                                                  Prefix='v0/')
+        for o in result.get('CommonPrefixes', []):
+            path = o.get('Prefix')
+            if MinioSpecUtils.path_to_dataset_name(path) is None:
+                continue
+            dataset_paths.add(path)
+        return dataset_paths
+
+    def get_file_stats(self, key) -> List[dict]:
+        stats = []
+
+        dataset_name = MinioSpecUtils.path_to_dataset_name(key)
+        basename, format = MinioSpecUtils.split_dataset(dataset_name)
+        s3_path = f's3a://{self.bucket_name}/{key}'
+
+        print("s3_path: ", s3_path)
+
+        df = format.spark_load(self.spark_session, s3_path)
+        df.printSchema()
+
+        for col in df.dtypes:
+            stat_vals = get_stats_by_column(basename, col[0], col[1], df, df.count())
+            stats.extend(stat_vals)
+        return stats
+
+    def extract(self) -> Any:
+        if not self._extract_iter:
+            self._extract_iter = self.get_stats_iter()
+        try:
+            stat_obj = next(self._extract_iter)
+            return self.column_stat_from_stat_obj(stat_obj)
+        except StopIteration:
+            return None
+
+    def column_stat_from_stat_obj(self, stat_obj) -> TableColumnStats:
+        stat = TableColumnStats(table_name=stat_obj['table_name'],
+                                col_name=stat_obj['column_name'],
+                                stat_name=stat_obj['stat_name'],
+                                stat_val='"' + stat_obj['stat_val'] + '"',
+                                # TODO: modify start/end epoch to reflect
+                                # date stats were collected
+                                start_epoch='123',
+                                end_epoch='123',
+                                db=self.db,
+                                cluster=self.bucket_name,
+                                schema=self.schema
+                                )
+        return stat
```

Comment:
This method doesn't seem to add much in terms of abstraction -- it converts a dict to a `TableColumnStats`, but it's not clear to me why `get_file_stats` doesn't just return a `List[TableColumnStats]` rather than the intermediary `dict`

Reply:
good point, the stats dict is old code. I'm refactoring it.